### ISSUE TITLE
cuckoo_fast_double performs potentially unsafe bitwise-copy of table elements

### ISSUE THREAD

**Issue by nickhutchinson** [efficient/libcuckoo#56]
cuckoo_fast_double() calls [`vector<Bucket>::resize()`](https://github.com/efficient/libcuckoo/blob/master/src/cuckoohash_map.hh#L1708) to double the size of the table. `Bucket` stores elements in a `std::aligned_storage<>` and doesn't define a copy or move constructor -- this means that calling resize() will perform a bitwise copy of table elements, instead of invoking their copy or move constructors.

This is usually okay, but not all types can be moved by bitwise copy, e.g.

```c++
class NonRelocatableType {
private:
  char buffer[1024];
  char * pointerToBuffer;
  ...
public:
  NonRelocatableType() : pointerToBuffer(buffer) {}
  ...
};
```
(Example taken from https://github.com/facebook/folly/blob/master/folly/docs/Traits.md)

I think `Bucket` should probably declare a move constructor that moves occupied slots using std::move.

**Comment by manugoyal** [comment]
Should be fixed in b8cb2f5dba60cec0a6209c67e79a45f6ee7697b9. Added a unit test based on your example in 503c394e66d00545167a40bbbd200476985d5be4.

**Comment by nickhutchinson** [comment]
Thanks!
